src/persist-client/src/impl/machine.rs

```diff
@@ -758,15 +761,86 @@ mod tests {
 
             let state = Arc::new(Mutex::new(DatadrivenState::default()));
             let cpu_heavy_runtime = Arc::new(CpuHeavyRuntime::new());
+            let write = Arc::new(Mutex::new(
+                client
+                    .open_writer::<String, (), u64, i64>(shard_id)
+                    .await
+                    .expect("invalid shard types"),
+            ));
 
             f.run_async(move |tc| {
                 let shard_id = shard_id.clone();
                 let client = client.clone();
                 let state = Arc::clone(&state);
                 let cpu_heavy_runtime = Arc::clone(&cpu_heavy_runtime);
+                let write = Arc::clone(&write);
                 async move {
                     let mut state = state.lock().await;
+
                     match tc.directive.as_str() {
+                        // Returns whether a blob contains a given batch. The batch
+                        // must contain only a single key.
+                        "blob-batch-exists" => {
+                            let input = get_arg(&tc.args, "input").expect("missing input");
+                            let batch = state.batches.get(input).expect("unknown batch").clone();
+                            let key = batch.keys.get(0).expect("unknown batch part");
+
+                            assert_eq!(batch.keys.len(), 1);
+
+                            let blob = client
+                                .blob
+                                .get(&key.complete(&shard_id))
+                                .await
+                                .expect("blob read failed");
+
+                            format!("{}\n", blob.is_some())
+                        }
+                        // Scans consensus and returns all states with their SeqNos
+                        // and which batches they reference
+                        "consensus-scan" => {
+                            let from =
+                                SeqNo(get_u64(&tc.args, "from_seqno").expect("missing from_seqno"));
+                            let res = client.consensus.scan(&shard_id.to_string(), from).await;
+
+                            let mut s = String::new();
+                            match res {
+                                Ok(states) => {
+                                    for persisted_state in states {
+                                        let persisted_state = {
+                                            use prost::Message;
```

Comment:
nit: move this use to the block at the top of `mod tests`

Reply:
ah so it conflicted with the separate decode call further below -- is there a way to say I want to specifically use prost::Message::decode?

Reply:
oh! `prost::Message::decode(&foo, &buf)` or `<foo as prost::Message>::decode(&buf)`. I prefer the former, usually